Transaction inclusion proofs enhancements

viquezclaudio opened the pull request:
When generating transaction inclusion proofs there are multiple cases that need to be considered in order to generate a proof that can be verified. From the requester side, transactions can be grouped into: 

1. Transactions that belong to finalized epochs (currently hashes_by_epoch )
2. Transactions that belong to an unfinished epoch, but have a checkpoint block that finalizes them
3. Transactions that belong to the current batch

To prove those transactions:
For 1 we essentially do what we are currently doing: use the election block of that epoch to prove those transactions
For 2 we should use the last_macro_block of the current epoch, and use the current MMR length to generate the proof
For 3 we use the last micro block number of the current batch, and the current MMR length to generate the proof

sisou committed:
Keep the accumulated transaction count of the current epoch in ChainInfo

This will be used to generate transaction inclusion proofs for transaction in previous batches in the current epoch, which require that the proof be generated with the length of the store at the batch's finalization checkpoint block.

viquezclaudio committed:
Handle multiple cases when requesting txns inclusion proofs.

- There are different scenarios that need to be considered when creating txn inclusion proofs
- A: The txns correspond to a finalized epoch
- B: The txns correspond to the current epoch from a finalized batch
- C: The txns correspond to the current batch

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -94,26 +95,59 @@ impl<N: Network> ConsensusProxy<N> {
+
+                    let checkpoint_head_number = self.blockchain.read().macro_head().block_number();
+
+                    let current_head_number = self.blockchain.read().head().block_number();
```
I suggest to use the same `read()` instance for all three blocks to have a consistent state.

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -94,26 +95,59 @@ impl<N: Network> ConsensusProxy<N> {
+                        if block_number <= election_head_number {
+                            // First Case: Transactions from finalized epochs
+                            hashes_by_block
+                                .entry(Policy::election_block_of(Policy::epoch_at(block_number)))
```
Isn't there a `election_after_block(height)` method or similar?

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -94,26 +95,59 @@ impl<N: Network> ConsensusProxy<N> {
 
-                    // Now we request proofs for each epoch and its hashes
-                    for (epoch, hashes) in hashes_by_epoch {
+                    // Now we request proofs for each epoch and its hashes, according to its classification
```
```suggestion
                    // Now we request proofs for each block and its hashes, according to their classification
```

sisou reviewed (approved):
Nice. You still have a clippy warning though to use `else if`.

viquezclaudio committed:
Simplify the computation of the txn count

The accumulated transaction count that is now included as part of ChainInfo is needed for generation txns proofs.
However, for the current epoch, all that is needed is the accumulated txn count since the beggining of the current epoch.
This is because the history tree operates on a per epoch basis (and not on a per batch basis)
So there is no need to compute, save and expose the accumulated txn count of each checkpoint block.

viquezclaudio commented:
@sisou I just pushed a new commit that might be relevant for you. Because I discussed with JD and Pascal, that for generating txns proofs of the current epoch all we need is the accumulated txn count up to the point where we want to generate the proof, and there is no need to keep track of that count every checkpoint block.

sisou commented:
I was not aware that it's possible to get a ChainInfo for a past block. In that case your change is correct, as for macro blocks both tx counters are/were the same.

jsdanielh commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -94,26 +95,58 @@ impl<N: Network> ConsensusProxy<N> {
-                    let mut hashes_by_epoch = HashMap::new();
+                    let blockchain = self.blockchain.read();
+
+                    // Group transaction hashes by the block number that proofs those transactions to reduce the number of requests
```
```suggestion
                    // Group transaction hashes by the block number that proves those transactions to reduce the number of requests
```

jsdanielh commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -94,26 +95,58 @@ impl<N: Network> ConsensusProxy<N> {
+
+                    let checkpoint_head_number = blockchain.macro_head().block_number();
+
+                    let current_head_number = blockchain.head().block_number();
```
I'd remove the extra lines
```suggestion
                    // This is the structure where we group transactions by their proving block number
                    let mut hashes_by_block = HashMap::new();
                    let election_head_number = blockchain.election_head().block_number();
                    let checkpoint_head_number = blockchain.macro_head().block_number();
                    let current_head_number = blockchain.head().block_number();
```

jsdanielh commented on primitives/block/src/block.rs:
```diff
@@ -232,8 +232,7 @@ impl Block {
     }
 
-    /// Return the number of transactions in the block. If the block is a Macro
-    /// block it just returns zero, since Macro blocks don't contain any transactions.
```
Why did we change it?

viquezclaudio replied:
I thought the function returns only the number of transactions of the block and there was no need to clarify the distinction between micro and macro

sisou replied:
I changed that comment actually in my commit, because as of this commit macro blocks now _do_ include transactions:
https://github.com/nimiq/core-rs-albatross/pull/1280/commits/26c8a848f500757d91f7452082d4636500d95e69#diff-aaa71b99350a00b080a331c7886311a206cfc2475c64d82748c711d9c4f10a04R242-R244

jsdanielh replied:
Ah, but we didn't then update the primitives properly: https://github.com/nimiq/core-rs-albatross/blob/albatross/primitives/block/src/block.rs#L221

viquezclaudio replied:
Ok sorry, Initially I was confused because I didn't remember that I made that change, but then I thought that maybe I just forgot about it :) 
Now it makes sense, I can update the primitives as well

viquezclaudio committed:
Applied PR comments

viquezclaudio committed:
Include logs for debugging and fixed lock usage

- Include some logs to debug the txns by address request
- Fix usage of the blockchain read lock while performing the request

jsdanielh committed:
Use the history tree length for generating a proof

Use the history tree length for generating a transaction proof since
this is what the history tree interface expects when generating a
proof for a previous state of the tree.
Also rename the `cum_ext_tx_count` field of `ChainInfo` to
`history_tree_len` and change the blockchain code to properly set it
based on the current length of the history tree.
